clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorRequestManager.java

```diff
@@ -0,0 +1,289 @@
+            if (durationOfOngoingDisconnect > this.rebalanceTimeoutMs)
+                log.debug("Consumer has been disconnected from the group coordinator for {}ms",
+                        durationOfOngoingDisconnect);
+        }
+    }
+
+    private void onSuccessfulResponse(
+            final FindCoordinatorResponseData.Coordinator coordinator,
+            final long currentTimeMS) {
+        // use MAX_VALUE - node.id as the coordinator id to allow separate connections
+        // for the coordinator in the underlying network client layer
+        int coordinatorConnectionId = Integer.MAX_VALUE - coordinator.nodeId();
+
+        this.coordinator = new Node(
+                coordinatorConnectionId,
+                coordinator.host(),
+                coordinator.port());
+        log.info("Discovered group coordinator {}", coordinator);
+        coordinatorRequestState.reset();
+        return;
+    }
+
+    private void handleFailedCoordinatorResponse(
+            final Exception exception,
+            final long currentTimeMs) {
+        coordinatorRequestState.updateLastFailedAttempt(currentTimeMs);
+        markCoordinatorUnknown("coordinator unavailable", currentTimeMs);
+
+        if (exception == Errors.GROUP_AUTHORIZATION_FAILED.exception()) {
+            log.debug("FindCoordinator request failed due to authorization error {}", exception.getMessage());
+            errorHandler.handle(GroupAuthorizationException.forGroupId(this.groupId));
+            return;
+        }
+
+        if (exception instanceof RetriableException) {
+            log.debug("FindCoordinator request failed due to retriable exception", exception);
+            return;
+        }
+
+        log.warn("FindCoordinator request failed due to fatal exception", exception);
+        errorHandler.handle(exception);
+    }
+
+    public void onResponse(final FindCoordinatorResponse response, final long currentTimeMs, final Throwable t) {
+        // handle Runtime exception
+        if (t != null) {
+            log.error("FindCoordinator request failed due to {}", t.getMessage());
+            return;
+        }
+
+        Optional<FindCoordinatorResponseData.Coordinator> coordinator = response.getCoordinatorByKey(this.groupId);
+        if (!coordinator.isPresent()) {
+            String msg = String.format("Coordinator not found for groupId: %s", this.groupId);
+            handleFailedCoordinatorResponse(new IllegalStateException(msg), currentTimeMs);
+            return;
+        }
+
+        FindCoordinatorResponseData.Coordinator node = coordinator.get();
+        if (node.errorCode() != Errors.NONE.code()) {
+            handleFailedCoordinatorResponse(Errors.forCode(node.errorCode()).exception(), currentTimeMs);
+            return;
+        }
+        onSuccessfulResponse(node, currentTimeMs);
+    }
+
+    public Node coordinator() {
+        return this.coordinator;
+    }
+
+    // Visible for testing
```

Comment:
The logic in this class seems general and not tied to the `FindCoordinator` API. Does it make sense to pull it out? Maybe we could call it `RequestTracker` or `RequestState` or something like that?

Reply:
Thanks, that make sense.